rad/src/lib.rs

```diff
@@ -26,56 +29,108 @@ pub mod types;
 
 pub type Result<T> = std::result::Result<T, RadError>;
 
-pub async fn try_data_request(request: &RADRequest) -> RadonReport<RadonTypes> {
+/// The return type of any method executing the entire life cycle of a data request.
+#[derive(Debug, Serialize)]
+pub struct RADRequestExecutionReport {
+    /// Vector of reports about retrieval of data sources.
+    pub retrieve: Vec<RadonReport<RadonTypes>>,
+    /// Report about aggregation of data sources.
+    pub aggregate: RadonReport<RadonTypes>,
+    /// Report about aggregation of reports (reveals, actually).
+    pub tally: RadonReport<RadonTypes>,
+}
+
+/// Executes a data request locally.
+/// The `inputs_injection` allows for disabling the actual retrieval of the data sources and
+/// the provided strings will be fed to the retrieval scripts instead. It is therefore expected that
+/// the length of `sources_injection` matches that of `request.retrieve`.
+pub async fn try_data_request(
+    request: &RADRequest,
+    settings: RadonScriptExecutionSettings,
+    inputs_injection: Option<&[&str]>,
+) -> RADRequestExecutionReport {
     let context = &mut ReportContext::default();
 
-    let retrieve_responses_fut = request
-        .retrieve
-        .iter()
-        .map(|retrieve| run_retrieval_report(retrieve));
+    let retrieve_responses = if let Some(inputs) = inputs_injection {
+        assert_eq!(inputs.len(), request.retrieve.len(), "Tried to locally run a data request with a number of injected sources different than the number of retrieval paths ({} != {})", inputs.len(), request.retrieve.len());
 
-    let retrieve_responses: Vec<RadonTypes> = futures::future::join_all(retrieve_responses_fut)
-        .await
+        let mut retrieve_responses = Vec::new();
+        for (retrieve, input) in request.retrieve.iter().zip(inputs.iter()) {
+            retrieve_responses.push(run_retrieval_with_data_report(
+                retrieve, input, context, settings,
+            ))
+        }
+
+        retrieve_responses
+    } else {
+        block_on(join_all(
+            request
+                .retrieve
+                .iter()
+                .map(|retrieve| run_retrieval_report(retrieve, settings))
+                .collect::<Vec<_>>(),
+        ))
+    };
+
+    let retrieval_reports: Vec<RadonReport<RadonTypes>> = retrieve_responses
         .into_iter()
         .map(|retrieve| {
-            retrieve
-                .unwrap_or_else(|error| RadonReport::from_result(Err(error), context))
-                .into_inner()
+            retrieve.unwrap_or_else(|error| RadonReport::from_result(Err(error), context))
         })
         .collect();
+    let retrieval_values: Vec<RadonTypes> = retrieval_reports
+        .iter()
+        .cloned()
+        .map(|report| report.into_inner())
```

Comment:
This is cloning the entire report, which may be expensive when the report includes the list of partial results. We just need to clone the final result:

```suggestion
        .map(|report| report.result.clone())
```